### DISCUSSION TITLE
How to add a progress ticker?

### DISCUSSION THREAD

**Discussion by Ra-Na** [amugofjava/anytime_podcast_player#110] [Q&A] [answered]
Dear community,

I would like to add a progress ticker that calls a custom function every second and passes in the current playtime of the episode. How can I do that? 

My understanding is that there are no less than 3 layers (!) handling the audio. 

just audio is the player <-> wrapped by audio service for background playback <-> audio bloc
 
What is the function then of audio bloc? 

I have set up android studio with flutter/dart and can run the App on a virtual device. 

This is what I considered so far:
- piggyback on the seek bar UI update. But I assume this will not work because when the UI is not visible the seek bar update is paused.
- derive what i want from the sleep timer. But the sleep timer seems to not have a progress ticker, but uses a built-in sleep function in just_audio. 

I would appreciate if someone could just give me a short list of steps that I need to do. 

I come from java/kotlin development for Android.

**Comment by amugofjava** [comment]
Hi @Ra-Na,

The role of the BLoCs is to handle communication between the UI and lower service layers, this is why I pause the ticker when the app goes into the background as there is no UI to update (and pausing it might save a bit of battery).

I'm not sure if your progress ticker needs to tick all the time or just when an episode is playing, but there are a couple of ways you could do this.

1. Create your own periodic stream to tick once every second and call your function or;
2. Hook into the [`positionStream`](https://pub.dev/documentation/just_audio/latest/just_audio/AudioPlayer/positionStream.html) offered by `just_audio` which updates between every 16ms - 200ms with the current playback position.

**Reply by Ra-Na** [reply, marked as answer]
Thank you very much. I will try the second approach.

**Comment by Ra-Na** [comment]
Hello Ben,

since a week now I am meditating over your code. It works like a charm. I have compared to some other open source podcast players for mobile devices, and it is most clean while including anything one may need. Thus I stick with it, although I am new to flutter/dart. But it requires a lot of effort to fully understand the flow of data. The code may benefit from a refactoring and a more detailed documentation. This might help to acquire more developers. 

Anyway, to the point:

I think the easiest way to include my function into the code is in `_onPositionUpdate` in `audio_player_service.dart`.
A simple `log.fine("TICKER $position");` shows that it works, I just have to replace the log command with my function.

Full disclosure: I am implementing a listener that skips ads, similar to SponsorBlock on Youtube. I have implemented the other missing pieces, namely a web service to report and look up ad segments and automatic detection of common text snippets in different audio files. The integration with a PodCatcher seems to be the hardest part.

I know that ad blocking is a controversial subject and understand and respect that others may disagree. For me, this project is mostly educational, but if something useful comes out, even better. 

Have a nice day,

Rainer

**Comment by amugofjava** [comment]
Hi @Ra-Na,

Thank you for the feedback. I'm always looking to improve documentation. Are there any particular areas you think need refactoring and what type of documentation do you think would improve the project - more code comments, more overall code about the architecture for example?

You could hook into the `_onUpdatePosition` method which ticks every 500ms. It does stop ticking when the UI is pushed into the background, but you could always remove those calls in the `resume()` and `suspend()` methods.

I hope you have a lot of fun developing your project.

**Reply by Ra-Na** [reply]
Thank you for the hint!

For the documentation, a flow chart that depicts which components interact in what way is probably best to give a quick overview. It takes some time to work this out from the code if one is new to the subject, but you can probably draw it from memory. And maybe a text file in each of the subfolders in lib describing the purpose each folder. 

Have a nice day,

Rainer

**Comment by Ra-Na** [comment]
Hello! Just wanted to inform you that I achieved what I wanted, but learned afterwards that ads are dynamically added to podcasts. A static timestamp is useless, one needs a dynamic ad detection. It is really hard to get this done. There is a discord channel where developers discuss this, but it seems pretty hopeless for now. The head of the "movement" is this guy: https://github.com/ajayyy, if you are interested.

Have a nice day,

Rainer

**Comment by amugofjava** [comment]
Hi. Yea, Dynamic Ad Insertion (DAI) is frustrating! It pushes timestamps out often rendering chapters and especially transcripts somewhat useless as they end up out of sync. More recently, disgruntled users who are annoyed by the adds are complaining to podcast developers and leaving poor app reviews as they think the apps are putting these adds in to make money, and don't realise they come from the hosting and ad companies.

Personally, I don't have any issue with advertising, but I find the lazy and poor implementations of ads frustrating. It could be done so much better.

**Reply by Ra-Na** [reply]
Unfortunately, there is no incentive for advertisers to implement better unless countermeasures are taken. I also noticed that transcripts and chapter timestamps are pretty useless due to DAI. I will meditate over alternative methods. A live transcript would be a step in the right direction, but I fear that would drain the batteries of mobile devices too fast.
